### PR TITLE
Expand System.Runtime.InteropServices.NFloat to support the APIs required by Xamarin

### LINKED ISSUE #63801: Extend `System.Runtime.InteropServices.NFloat` so it can replace the Xamarin in-box `nfloat` for .NET 7
### Rationale

In https://github.com/dotnet/runtime/issues/13788 we exposed some new interchange types `CLong`, `CULong`, and `NFloat` to help ensure interop with underlying runtimes could be done in a stable and cross platform manner.

We initiallizy decided that these shoulld not expose more complex support such as operators, parsing, or formatting. However, this decision means that `Xamarin` must continue maintaining their own version of `ObjCRuntime.nfloat`.

As per https://github.com/xamarin/xamarin-macios/issues/13087#issuecomment-1012968826, exposing this additional functionality on `NFloat` would allow them to remove their custom type and normalize on the "built-in" functionality instead.

### Proposal

```csharp
namespace System.Runtime.InteropServices;

public readonly struct NFloat
    : IComparable,
      IComparable<NFloat>,
      IConvertible,         // We've explicitly decided against exposing this in other cases
      IEquatable<NFloat>,   // Already exposed by 13788
      IFormattable
{
    // Xamarin exposes these as `static readonly` but that is "less optimizable" and not our normal convention
    public static NFloat Epsilon { get; }
    public static NFloat MaxValue { get; }
    public static NFloat MinValue { get; }
    public static NFloat NaN { get; }
    public static NFloat NegativeInfinity { get; }
    public static NFloat PositiveInfinity { get; }
    public static NFloat Size { get; }

    public static bool IsInfinity(NFloat value);
    public static bool IsNaN(NFloat value);
    public static bool IsNegativeInfinity(NFloat value);
    public static bool IsPositiveInfinity(NFloat value);

    // Parsing APIs, would match IParseable<NFloat> and IBinaryFloatingPoint<NFloat> if exposed
    public static NFloat Parse(string s);
    public static NFloat Parse(string s, NumberStyles style);
    public static NFloat Parse(string s, IFormatProvider? provider);
    public static NFloat Parse(string s, NumberStyles style, IFormatProvider? provider);
    public static bool TryParse([NotNullWhen(true)] string? s, out NFloat result);
    public static bool TryParse([NotNullWhen(true)] string? s, NumberStyles style, IFormatProvider? provider, out NFloat result);

    // Arithmetic Operators
    public static NFloat operator +(NFloat left, NFloat right);
    public static NFloat operator --(NFloat value);
    public static NFloat operator /(NFloat left, NFloat right);
    public static NFloat operator ++(NFloat value);
    public static NFloat operator %(NFloat left, NFloat right);
    public static NFloat operator *(NFloat left, NFloat right);
    public static NFloat operator -(NFloat left, NFloat right);
    public static NFloat operator +(NFloat value);
    public static NFloat operator -(NFloat value);

    // Comparison Operators
    public static bool operator ==(NFloat left, NFloat right);
    public static bool operator >(NFloat left, NFloat right);
    public static bool operator >=(NFloat left, NFloat right);
    public static bool operator !=(NFloat left, NFloat right);
    public static bool operator <(NFloat left, NFloat right);
    public static bool operator <=(NFloat left, NFloat right);

    // Explicit * to NFloat conversions
    public static explicit operator NFloat(decimal value);
    public static explicit operator NFloat(double value);
    public static explicit operator NFloat(nint value);     // Explicit is inconsistent with int and long
    public static explicit operator NFloat(nuint value);    // Explicit is inconsistent with uint and ulong, Xamarin is missing this one

    // Explicit NFloat to * conversions
    public static explicit operator byte(NFloat value);
    public static explicit operator char(NFloat value);
    public static explicit operator decimal(NFloat value);
    public static explicit operator short(NFloat value);
    public static explicit operator int(NFloat value);
    public static explicit operator long(NFloat value);
    public static explicit operator nint(NFloat value);
    public static explicit operator sbyte(NFloat value);
    public static explicit operator float(NFloat value);
    public static explicit operator ushort(NFloat value);
    public static explicit operator uint(NFloat value);
    public static explicit operator ulong(NFloat value);
    public static explicit operator nuint(NFloat value);   // Xamarin is missing this one

    // Implicit * to NFloat conversions
    public static implicit operator NFloat(byte value);
    public static implicit operator NFloat(char value);
    public static implicit operator NFloat(short value);
    public static implicit operator NFloat(int value);
    public static implicit operator NFloat(long value);
    public static implicit operator NFloat(sbyte value);
    public static implicit operator NFloat(float value);
    public static implicit operator NFloat(ushort value);
    public static implicit operator NFloat(uint value);
    public static implicit operator NFloat(ulong value);

    // Implicit NFloat to * conversions
    public static implicit operator double(NFloat value);

    // From IComparable and IComparable<NFloat>
    public bool CompareTo(NFloat other);
    public bool CompareTo(object other);

    // From IConvertible, if we decide to support
    public TypeCode GetTypeCode();
    bool IConvertible.ToBoolean(IFormatProvider? provider);
    byte IConvertible.ToByte(IFormatProvider? provider);
    char IConvertible.ToChar(IFormatProvider? provider);
    DateTime IConvertible.ToDateTime(IFormatProvider? provider);
    decimal IConvertible.ToDecimal(IFormatProvider? provider);
    double IConvertible.ToDouble(IFormatProvider? provider);
    short IConvertible.ToInt16(IFormatProvider? provider);
    int IConvertible.ToInt32(IFormatProvider? provider);
    long IConvertible.ToInt64(IFormatProvider? provider);
    sbyte IConvertible.ToSByte(IFormatProvider? provider);
    float IConvertible.ToSingle(IFormatProvider? provider);
    string IConvertible.ToString(IFormatProvider? provider);
    object IConvertible.ToType(Type conversionType, IFormatProvider? provider);
    ushort IConvertible.ToUInt16(IFormatProvider? provider);
    uint IConvertible.ToUInt32(IFormatProvider? provider);
    ulong IConvertible.ToUInt64(IFormatProvider? provider);

    // From object and IEquatable<NFloat>, already exposed
    public bool Equals(NFloat other);
    public bool Equals(object other);

    // From object and IFormattable
    public string ToString();
    public string ToString(IFormatProvider? provider);
    public string ToString(string? format);
    public string ToString(string? format, IFormatProvider? provider);
}
```

Additionally, it may be worth making this consistent with `Half`/`Single`/`Double` with regards to the generic math work that is being done. This would additionally implement `IBinaryFloatingPoint<float>` and `IMinMaxValue<float>`.

### PR DESCRIPTION
**Pull request by tannergooding**
This resolves https://github.com/dotnet/runtime/issues/63801.

Tests still need to be added.

### CODE REVIEW

**Comment by tannergooding** [comment]
CC. @rolfbjarne, @marek-safar, @jeffhandley

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NFloat.cs [hunk 1]
```diff
@@ -13,65 +17,691 @@
...
-    /// or <c>double</c> on 64-bit platforms, such as the CGFloat type in libraries provided by Apple.
-    /// </summary>
+    /// <summary>Defines an immutable value type that represents a floating type that has the same size as the native integer size.</summary>
+    /// <remarks>It is meant to be used as an exchange type at the managed/unmanaged boundary to accurately represent in managed code unmanaged APIs that use a type alias for C or C++'s <c>float</c> on 32-bit platforms or <c>double</c> on 64-bit platforms, such as the CGFloat type in libraries provided by Apple.</remarks>
     [Intrinsic]
-    public readonly struct NFloat : IEquatable<NFloat>
+    public readonly struct NFloat
+        : IComparable,
+          IComparable<NFloat>,
+          IEquatable<NFloat>,
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NFloat.cs:26]: What about all the static numerical interfaces?  Is that separate?

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NFloat.cs:26]: Yes, going to cover that when everything else is fixed up to minimize overall churn.

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NFloat.cs [hunk 2]
```diff
@@ -13,65 +17,691 @@
...
+        /// <returns><c>true</c> if <paramref name="left" /> is less than <paramref name="right" />; otherwise, <c>false</c>.</returns>
+        [NonVersionable]
+        public static bool operator <(NFloat left, NFloat right) => left._value < right._value;
+
+        /// <summary>Compares two values to determine which is less or equal.</summary>
+        /// <param name="left">The value to compare with <paramref name="right" />.</param>
+        /// <param name="right">The value to compare with <paramref name="left" />.</param>
+        /// <returns><c>true</c> if <paramref name="left" /> is less than or equal to <paramref name="right" />; otherwise, <c>false</c>.</returns>
+        [NonVersionable]
+        public static bool operator <=(NFloat left, NFloat right) => left._value >= right._value;
```
Comment (rolfbjarne) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NFloat.cs:202]:
```suggestion
        public static bool operator <=(NFloat left, NFloat right) => left._value <= right._value;
```

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NFloat.cs [hunk 3]
```diff
@@ -13,65 +17,691 @@
...
+        /// <returns><c>true</c> if <paramref name="left" /> is greater than <paramref name="right" />; otherwise, <c>false</c>.</returns>
+        [NonVersionable]
+        public static bool operator >(NFloat left, NFloat right) => left._value > right._value;
+
+        /// <summary>Compares two values to determine which is greater or equal.</summary>
+        /// <param name="left">The value to compare with <paramref name="right" />.</param>
+        /// <param name="right">The value to compare with <paramref name="left" />.</param>
+        /// <returns><c>true</c> if <paramref name="left" /> is greater than or equal to <paramref name="right" />; otherwise, <c>false</c>.</returns>
+        [NonVersionable]
+        public static bool operator >=(NFloat left, NFloat right) => left._value <= right._value;
```
Comment (rolfbjarne) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NFloat.cs:216]:
```suggestion
        public static bool operator >=(NFloat left, NFloat right) => left._value >= right._value;
```

**Review by rolfbjarne — CHANGES_REQUESTED** [review]
Looks like two operators are reversed?

**Review by rolfbjarne — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NFloat.cs [hunk 4]
```diff
@@ -13,65 +17,691 @@
...
+        public static bool operator ==(NFloat left, NFloat right) => left._value == right._value;
+
+        /// <summary>Compares two values to determine inequality.</summary>
+        /// <param name="left">The value to compare with <paramref name="right" />.</param>
+        /// <param name="right">The value to compare with <paramref name="left" />.</param>
+        /// <returns><c>true</c> if <paramref name="left" /> is not equal to <paramref name="right" />; otherwise, <c>false</c>.</returns>
+        [NonVersionable]
+        public static bool operator !=(NFloat left, NFloat right) => left._value != right._value;
+
+        /// <summary>Compares two values to determine which is less.</summary>
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NFloat.cs:190]: Nit: This description seems a bit misleading, as it tells you whether left is less than right but not whether right is strictly less then left.

Consider adopting the existing doc terminology that float/double use, e.g. "Returns a value that indicates whether a specified Double value is less than or equal to another specified Double value."

**Review by stephentoub — APPROVED** [review]
Gave it a quick skim and generally LGTM.

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NFloat.cs (hunk 4 above)
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NFloat.cs:190]: > Consider adopting the existing doc terminology

I always forget that we have APIs exposed here that are functionally unavailable and will never be called 👍 

I'll update this in a follow up PR as I need to also update `IComparisonOperators` and a few other places accordingly.

**Pull request merged**